File: geekbang/algorithm/source-code/1/lc206 反转链表.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode now = head;
        ListNode last = null;
        // 遍历
        while (now != null) {
            // 备份now.next
            ListNode next = now.next;
            // 操作：now指向last
            now.next = last;
            // 往后移动
            last = now;
            now = next;
        }
        return last;
    }
}

File: geekbang/algorithm/source-code/1/lc206 反转链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        now = head
        last = None
        # 遍历
        while now:
            # 备份now.next
            next = now.next
            # 操作：now指向last
            now.next = last
            # 往后移动
            last = now
            now = next
        return last